src/include/sof/common.h

```diff
@@ -177,6 +177,46 @@
 #define IS_ENABLED_STEP_3(ignore, value, ...) (!!(value))
 #endif
 
+#define SOF_AUTO_HIFI -1
+
+/* Generic, no HIFI. This could be anything but it is zero for
+ * user-friendliness. Unfortunately, an accidentally undefined
+ * CONFIG_${comp}_HIFI will silently be converted to 0 and will not be
+ * reported. To catch such issues, simply set this to -2 temporarily.
+ */
+#define SOF_NO_HIFI 0
+
+/* True if CONFIG_${comp}_HIFI has been explicitly configured to value 'v'.
+ *
+ * We unfortunately cannot used `defined()` here:
+ *  - https://bugs.webkit.org/show_bug.cgi?id=167643
+ *  - https://lists.llvm.org/pipermail/cfe-commits/Week-of-Mon-20160118/147239.html
+ */
+#define SOF_CONFIG_HIFI(v, comp) ((v) == (CONFIG_ ## comp ## _HIFI))
+
+/* True if either explicitly specified in Kconfig as above, _or_
+ *   it is the max HIFI available in the HAL.
+ */
+#define SOF_USE_HIFI(v, comp) SOF_CONFIG_HIFI(v, comp) || \
+	(SOF_CONFIG_HIFI(SOF_AUTO_HIFI, comp) && SOF_MAX_XCHAL_HIFI == (v))
+
+#ifndef __XCC__ // Cadence toolchains
+#  define SOF_MAX_XCHAL_HIFI SOF_NO_HIFI
+#else
+#  include <xtensa/config/core-isa.h>
+// Maybe we could make this fully generic (and less readable!) using
```

Comment:
I am not sure this is best way to go, with this, module HIFI config still depend on core-isa, this is centuralized. 

The logic is more complex and with more dependency, like __XCC__, more complex SOF_USE_HIFI macro definition, if expand this to other dsp platform(non-xtensa and X86), this part will need change again and hard to make changes.

Reply:
> module HIFI config still depend on core-isa, this is centuralized

You haven't summarized why these are bad.

We can duplicate `core-isa.h` in various places and I'm looking at options but `core-isa.h` will always remain the origin, true reference for HIFI facts.

> with more dependency, like `__XCC__`

`__XCC__` is the only reliable way to detect Cadence toolchains and it's what Zephyr uses.
https://github.com/thesofproject/sof/pull/8682#discussion_r1446760651

Which other dependency? 


> The logic is more complex [...] and hard to make changes.

If you find this logic complex then I'm afraid nothing can be done for you.